hwdef: Add AP_HW_JHEMCUGF16F405 board config

alex-pokydin opened the pull request:
Add support for hardware
https://www.jhemcu.com/e_productshow/?76-JHEMCU-GF16-BMI-stack-F405-Flight-control-13A-ESC-76.html

Hwurzburg reviewed (changes requested):
needs readme file with information like this one:https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/SpeedyBeeF405WING/Readme.md

once created,I will review again

alex-pokydin commented:
> needs readme file with information like this one:https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/SpeedyBeeF405WING/Readme.md
> 
> once created,I will review again

@Hwurzburg , done, please, review the PR,
thanks!

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|:-|:-|:-|
+|SERIAL0|COMPUTER|USB|
+|SERIAL1|RX1/TX1|USART1 (DMA) - Serial RC input (DMA) |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User, Optic Flow |
```
```suggestion
|SERIAL2|TX2/RX2|USART2 (DMA RX only) - User |
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|SERIAL0|COMPUTER|USB|
+|SERIAL1|RX1/TX1|USART1 (DMA) - Serial RC input (DMA) |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User, Optic Flow |
+|SERIAL3|TX3/RX3|USART3 (DMA) - User, GPS |
```
```suggestion
|SERIAL3|TX3/RX3|USART3 (DMA TX only) - User, GPS |
```

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|SERIAL1|RX1/TX1|USART1 (DMA) - Serial RC input (DMA) |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User, Optic Flow |
+|SERIAL3|TX3/RX3|USART3 (DMA) - User, GPS |
+|SERIAL4|TX4/RX4|UART4 (DMA) - User, VTX |
```
```suggestion
|SERIAL4|TX4/RX4|UART4 (DMA TX only) - User, VTX |
```

if you really intend this to be for VTX control, either as SmartAudio or DJI or Display Port, it will need a hwdef define to set the protocol...otherwise its GPS by default

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|Name|Pin|Function|
+|:-|:-|:-|
+|SERIAL0|COMPUTER|USB|
+|SERIAL1|RX1/TX1|USART1 (DMA) - Serial RC input (DMA) |
```
I would add a define to set the default to RCIN protocol in hwdef

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|SERIAL2|TX2/RX2|USART2 (DMA) - User, Optic Flow |
+|SERIAL3|TX3/RX3|USART3 (DMA) - User, GPS |
+|SERIAL4|TX4/RX4|UART4 (DMA) - User, VTX |
+|SERIAL6|TX6/RX6|USART6 -  User, Telemetry|
```
```suggestion
|SERIAL6|TX6/RX6|USART6 -  User|
```

Hwurzburg replied:
still there telem is no the default protocol for usart6....I am approving anyway and will fix when I do the wiki page

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+
+## RC Input
+ 
+RC input is configured on SERIAL1 (USART1), configured for CRSF with `SERIAL1_PROTOCOL 23`.
```
SBUS inout must be on SBUS pin, all others can use RX1, CRSF/ELRS/SRXL2 must also use TX pin

Hwurzburg replied:
sorry for my typo

Hwurzburg reviewed (changes requested):
also, should have only two commits
one for the bootloaders in Tools and one for hwdef in AP_HAL_ChibiOS

alex-pokydin commented:
@Hwurzburg , thank you for your corrections.
I do my best to fix/cleanup all points
Please, check if all changes are correct
Thank you!

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|SERIAL1|RX1/TX1|USART1 (DMA) - RC Input/Output |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User |
+|SERIAL3|TX3/RX3|USART3 (DMA) - User |
+|SERIAL4|TX4/RX4|UART4 (DMA) - User |
```
```suggestion
|SERIAL4|TX4/RX4|UART4 (DMA TX only) - GPS2 |
```

Hwurzburg replied:
and you did not adddress my previous comment...based on your wiring diagram you expect users to use this port for VTX control...you should set the protocol in the hwdef to DEFAULT_SERIAL4_PROTOCOL SerialProtocol_Tramp or _SmartAudio

alex-pokydin replied:
can we keep the port just as a general purpose port, so users will decide how they will use it?

Hwurzburg replied:
you can but BF users using your wiring diagram will have a harder time transitioning to AP

Hwurzburg commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+
+## RC Input
+ 
+SBUS inout must be on SBUS pin, all others can use RX1, CRSF/ELRS/SRXL2 must also use TX pin
```
```suggestion
SBUS input must be on SBUS pin, all others can use RX1, CRSF/ELRS/SRXL2 must also use TX pin
```

Hwurzburg reviewed (changes requested):
a few things were not addressed

alex-pokydin commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|:-|:-|:-|
+|SERIAL0|COMPUTER|USB|
+|SERIAL1|RX1/TX1|USART1 (DMA) - RC Input/Output |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User |
```
@Hwurzburg , could you, please, explain the reason of this change?

Hwurzburg replied:
see below

alex-pokydin commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/readme.md:
```diff
@@ -0,0 +1,111 @@
+|SERIAL0|COMPUTER|USB|
+|SERIAL1|RX1/TX1|USART1 (DMA) - RC Input/Output |
+|SERIAL2|TX2/RX2|USART2 (DMA) - User |
+|SERIAL3|TX3/RX3|USART3 (DMA) - User |
```
could you, please, explain the reason of this change?

Hwurzburg replied:
because the third and fourth uart protocols are set as GSP by default....and I checked the DMA assignments in the build configuration after waf configure and only those pins got DMA channels assigned

alex-pokydin commented:
> a few things were not addressed

@Hwurzburg , sorry again)
I've just added VTX to hwdef and added your corrections to uarts
please, check if it is ok now
Thank you!

Hwurzburg reviewed (approved):
looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

alex-pokydin commented:
> looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

Thank you!
BIDIR - it depends on ESC?

Hwurzburg commented:
yes....but a user may change from the bundled one or you may decide to upgrade the one you supply...if you dont include the capabiliiy, you cant have anyone migrate in the future without a firmware change...it has no downside to to include it...IF it can be done without removing DMAs from UART1 which you need for CRSF...your choice

alex-pokydin commented:
> looks good...you might want to contact AndyPiper and see if you can get BIDIR setup on at least the first four outputs...your choice

@andyp1per , could you, please, help me with BIDIR ? How can I check it and set it up?
Thank you!

Hwurzburg commented:
@alex-pokydin I have again asked @andyp1per to look at this PR to suggest how to add bidir

alex-pokydin commented:
@Hwurzburg , lets keep it as it is for now and I'll add bidir in the next iteration if it is possible
what do you think?

Hwurzburg commented:
your call I''l add a devcall label

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef-bl.dat:
```diff
@@ -0,0 +1,43 @@
+MCU STM32F4xx STM32F405xx
+
+# board ID for firmware load
+APJ_BOARD_ID 1081
```
can use AP_HW_JHEMCUGF16F405 instead of the numebr

alex-pokydin replied:
fixed

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/hwdef.dat:
```diff
@@ -0,0 +1,153 @@
+# board ID for firmware load
+# https://github.com/ArduPilot/ardupilot/blob/master/Tools/AP_Bootloader/board_types.txt
+# AP_HW_JHEMCUGF16F405
+APJ_BOARD_ID 1081
```
use AP_HW_JHEMCUGF16F405 not the number

alex-pokydin replied:
fixed

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/JHEM_JHEF405/JHEF-JHEF405PRO.config:
```diff
@@ -0,0 +1,170 @@
+# Betaflight / STM32F405 (S405) 4.2.0 Nov 22 2020 / 18:38:23 (afdac08b3) MSP API: 1.43
+
+board_name JHEF405
```
i don't think we want to include the betaflight config in our tree

alex-pokydin replied:
fixed

tridge reviewed (changes requested):
the commit messages need fixing, so "hwdef: added JHEMCUGF16F405"
and "Tools: added JHEMCUGF16F405 bootloader"

tridge commented:
@alex-pokydin if you have trouble with the git commit fixups then I can do it for you
ping me in code_review channel

alex-pokydin commented:
> @alex-pokydin if you have trouble with the git commit fixups then I can do it for you ping me in code_review channel

Hello @tridge , 
fixed, please, check it
thanks!
